gen/version.go

```diff
@@ -0,0 +1,185 @@
+// the License.
+
+//go:generate go run version.go
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"text/template"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/plumbing/storer"
+	"golang.org/x/mod/modfile"
+)
+
+const versionFileTemplate = `// Copyright 2021 Alvalor S.A.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package configuration
+
+const (
+	RosettaVersion    = "{{ .RosettaVersion }}"
+	NodeVersion       = "{{ .NodeVersion }}"
+	MiddlewareVersion = "{{ .MiddlewareVersion }}"
+)
+`
+
+func main() {
+	const rosettaVersion = "v1.4.10"
+	const rosettaVersionFilePath = "../rosetta/configuration/version.go"
+
+	fmt.Println("Using rosetta version", rosettaVersion)
+
+	nodeVersion, err := NodeVersion()
+	if err != nil {
+		log.Fatalf("could not compute node version: %v", err)
+	}
+
+	fmt.Println("Found node version", nodeVersion)
+
+	middlewareVersion, err := MiddlewareVersion()
+	if err != nil {
+		log.Fatalf("could not compute middleware version: %v", err)
+	}
+
+	fmt.Println("Found middleware version", middlewareVersion)
+
+	tmpl := template.Must(template.New("version.go").Parse(versionFileTemplate))
+
+	versionFile, err := os.Create(rosettaVersionFilePath)
+	if err != nil {
+		log.Fatalf("could not open version file: %v", err)
+	}
+
+	args := struct {
+		RosettaVersion    string
+		NodeVersion       string
+		MiddlewareVersion string
+	}{
+		RosettaVersion:    rosettaVersion,
+		NodeVersion:       nodeVersion,
+		MiddlewareVersion: middlewareVersion,
+	}
+
+	err = tmpl.Execute(versionFile, args)
+	if err != nil {
+		log.Fatalf("could not execute template: %v", err)
+	}
+}
+
+func NodeVersion() (string, error) {
+	// Fetch Node version from the go.mod file.
+	gomod, err := os.ReadFile("../go.mod")
+	if err != nil {
+		return "", fmt.Errorf("could not read go mod file: %v", err)
+	}
+
+	modfile, err := modfile.Parse("go.mod", gomod, nil)
+	if err != nil {
+		return "", fmt.Errorf("could not parse go mod file: %v", err)
+	}
+
+	var nodeVersion string
+	for _, module := range modfile.Require {
+		if module.Mod.Path == "github.com/onflow/flow-go" {
+			nodeVersion = module.Mod.Version
+			break
+		}
+	}
+
+	if nodeVersion == "" {
+		return "", fmt.Errorf("could not find github.com/onflow/flow-go dependency in go mod file")
+	}
+
+	return nodeVersion, nil
```

Comment:
If we just return upon success inside the loop, this code path just turns into a single error return.